File: lua/copilot/json_rpc_client.lua
--- JSON-RPC 2.0 client for Content-Length header framed stdio transport.
-- Reads from a file handle (stdout of the CLI process) and writes to another
-- (stdin of the CLI process). Uses coroutines for cooperative async I/O.

local cjson = require("cjson")

-- ---------------------------------------------------------------------------
-- UUID v4 generator (no external dependency)
-- ---------------------------------------------------------------------------

local random = math.random

--- Generate a UUID v4 string.
-- @return string A UUID in the form xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx
local function generate_uuid()
    -- Seed once per process
    local template = "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx"
    return (string.gsub(template, "[xy]", function(c)
        local v = (c == "x") and random(0, 0xf) or random(8, 0xb)
        return string.format("%x", v)
    end))
end

-- ---------------------------------------------------------------------------
-- JsonRpcClient class
-- ---------------------------------------------------------------------------

local JsonRpcClient = {}
JsonRpcClient.__index = JsonRpcClient

--- Create a new JSON-RPC 2.0 client.
-- @param write_handle file-like object with a :write() method (stdin pipe)
-- @param read_handle  file-like object with a :read() method  (stdout pipe)
-- @return JsonRpcClient
function JsonRpcClient.new(write_handle, read_handle)
    local self = setmetatable({}, JsonRpcClient)
    self._writer           = write_handle
    self._reader           = read_handle
    self._pending_requests = {}       -- id -> {co=coroutine, result=nil}
    self._request_handlers = {}       -- method -> function(params) -> result, err
    self._running          = false
    self._stopped          = false
    return self
end

--- Register a handler for incoming requests / notifications from the server.
-- @param method string  The JSON-RPC method name
-- @param handler function(params) -> result_table|nil, error_table|nil
function JsonRpcClient:set_request_handler(method, handler)
    self._request_handlers[method] = handler
end

--- Send a JSON-RPC 2.0 request and return the result.
-- MUST be called from inside the read loop (or a coroutine yielding back to it).
-- For the simple synchronous case, use request_sync() which drives the read loop.
-- @param method string
-- @param params table
-- @return table|nil result
-- @return string|nil error message
function JsonRpcClient:request(method, params)
    local id = generate_uuid()

    local request = {
        jsonrpc = "2.0",
        id      = id,
        method  = method,
        params  = params or {},
    }

    local ok, err = self:_send_message(request)
    if not ok then
        return nil, "failed to send request: " .. tostring(err)
    end

    -- Create an entry so the read loop can fill in the response
    self._pending_requests[id] = { result = nil, error = nil, done = false }

    -- If we are inside a coroutine (not the main thread), yield until the response arrives
    local co, is_main = coroutine.running()
    if co and not is_main then
        self._pending_requests[id].co = co
        -- Yield control back to the read loop driver
        local res_result, res_err = coroutine.yield()
        return res_result, res_err
    end

    -- Otherwise, spin the read loop inline (synchronous fallback)
    while not self._pending_requests[id].done do
        local read_ok, read_err = self:_read_one_message()
        if not read_ok then
            self._pending_requests[id] = nil
            return nil, "read error while waiting for response: " .. tostring(read_err)
        end
    end

    local entry = self._pending_requests[id]
    self._pending_requests[id] = nil
    if entry.error then
        return nil, entry.error
    end
    return entry.result, nil
end

--- Send a JSON-RPC 2.0 notification (fire and forget, no response expected).
-- @param method string
-- @param params table
-- @return boolean success
-- @return string|nil error message
function JsonRpcClient:notify(method, params)
    local notification = {
        jsonrpc = "2.0",
        method  = method,
        params  = params or {},
    }
    return self:_send_message(notification)
end

--- Run the read loop. Blocks until the connection is closed or stop() is called.
-- Dispatches incoming responses to pending requests and incoming requests/notifications
-- to registered handlers.
function JsonRpcClient:run()
    self._running = true
    while self._running do
        local ok, err = self:_read_one_message()
        if not ok then
            if self._running then
                io.stderr:write("JsonRpcClient read error: " .. tostring(err) .. "\n")
            end
            break
        end
    end
    self._running = false

    -- Wake up any pending requests with an error
    for id, entry in pairs(self._pending_requests) do
        entry.done = true
        entry.error = "client stopped"
        if entry.co then
            local ok, err = coroutine.resume(entry.co, nil, "client stopped")
            if not ok then
                io.stderr:write("coroutine resume error: " .. tostring(err) .. "\n")
            end
        end
    end
    self._pending_requests = {}
end

--- Stop the read loop.
function JsonRpcClient:stop()
    self._running = false
    self._stopped = true
end

--- Check whether the client is currently running.
-- @return boolean
function JsonRpcClient:is_running()
    return self._running
end

-- ---------------------------------------------------------------------------
-- Internal: wire protocol
-- ---------------------------------------------------------------------------

--- Write a JSON-RPC message with Content-Length framing.
-- @param message table  The message to encode and send
-- @return boolean success
-- @return string|nil error
function JsonRpcClient:_send_message(message)
    local ok_enc, data = pcall(cjson.encode, message)
    if not ok_enc then
        return false, "JSON encode error: " .. tostring(data)
    end

    local frame = string.format("Content-Length: %d\r\n\r\n%s", #data, data)
    local ok, err = pcall(function()
        self._writer:write(frame)
        self._writer:flush()
    end)
    if not ok then
        return false, "write error: " .. tostring(err)
    end
    return true, nil
end

--- Read exactly n bytes from the reader.
-- @param n number
-- @return string|nil data
-- @return string|nil error
function JsonRpcClient:_read_bytes(n)
    local data = self._reader:read(n)
    if data == nil then
        return nil, "EOF"
    end
    if #data < n then
        -- Some Lua I/O implementations may return partial reads; accumulate.
        local parts = { data }
        local got = #data
        while got < n do
            local chunk = self._reader:read(n - got)
            if chunk == nil then
                return nil, "EOF during read"
            end
            parts[#parts + 1] = chunk
            got = got + #chunk
        end
        data = table.concat(parts)
    end
    return data, nil
end

--- Read one line (up to and including \n) from the reader.
-- @return string|nil line
-- @return string|nil error
function JsonRpcClient:_read_line()
    local line = self._reader:read("*l")
    if line == nil then
        return nil, "EOF"
    end
    return line, nil
end

--- Read one Content-Length framed JSON-RPC message and dispatch it.
-- @return boolean success
-- @return string|nil error
function JsonRpcClient:_read_one_message()
    -- Read headers until blank line
    local content_length = 0
    while true do
        local line, err = self:_read_line()
        if not line then
            return false, err or "EOF"
        end

        -- Strip trailing \r if present
        line = line:gsub("\r$", "")

        -- Blank line signals end of headers
        if line == "" then
            break
        end

        -- Parse Content-Length header
        local length = line:match("^Content%-Length:%s*(%d+)")
        if length then
            content_length = tonumber(length)
        end
    end

    if content_length == 0 then
        -- Empty frame, skip
        return true, nil
    end

    -- Read the message body
    local body, err = self:_read_bytes(content_length)
    if not body then
        return false, "failed to read body: " .. tostring(err)
    end

    -- Decode JSON
    local ok_dec, msg = pcall(cjson.decode, body)
    if not ok_dec then
        return false, "JSON decode error: " .. tostring(msg)
    end

    -- Dispatch based on whether it is a request/notification or a response
    if msg.method and msg.method ~= "" then
        self:_handle_incoming_request(msg)
    elseif msg.id then
        self:_handle_response(msg)
    end

    return true, nil
end

--- Handle an incoming response to one of our pending requests.
-- @param response table
function JsonRpcClient:_handle_response(response)
    local id = response.id
    local entry = self._pending_requests[id]
    if not entry then
        return
    end

    if response.error then
        local err_msg = string.format(
            "JSON-RPC Error %s: %s",
            tostring(response.error.code or "?"),
            tostring(response.error.message or "unknown")
        )
        entry.error = err_msg
    else
        entry.result = response.result
    end
    entry.done = true

    -- If a coroutine is waiting, resume it
    if entry.co then
        self._pending_requests[id] = nil
        local ok, resume_err = coroutine.resume(entry.co, entry.result, entry.error)
        if not ok then
            io.stderr:write("coroutine resume error: " .. tostring(resume_err) .. "\n")
        end
    end
end

--- Handle an incoming request or notification from the server.
-- @param request table with fields: jsonrpc, id (may be nil), method, params
function JsonRpcClient:_handle_incoming_request(request)
    local handler = self._request_handlers[request.method]
    local is_call = (request.id ~= nil)

    if not handler then
        if is_call then
            self:_send_error_response(
                request.id,
                -32601,
                "Method not found: " .. tostring(request.method)
            )
        end
        return
    end

    -- Execute the handler
    local ok, result_or_err, rpc_err = pcall(handler, request.params)

    if not ok then
        -- Handler threw an error
        if is_call then
            self:_send_error_response(
                request.id,
                -32603,
                "Handler error: " .. tostring(result_or_err)
            )
        end
        return
    end

    if is_call then
        if rpc_err then
            self:_send_error_response(
                request.id,
                rpc_err.code or -32603,
                rpc_err.message or "internal error"
            )
        else
            self:_send_result_response(request.id, result_or_err)
        end
    end
end

--- Send a successful JSON-RPC response.
-- @param id string|number  The request ID
-- @param result table|nil  The result payload
function JsonRpcClient:_send_result_response(id, result)
    local response = {
        jsonrpc = "2.0",
        id      = id,
        result  = result or cjson.null,
    }
    self:_send_message(response)
end

--- Send an error JSON-RPC response.
-- @param id string|number  The request ID
-- @param code number       Error code
-- @param message string    Error message
function JsonRpcClient:_send_error_response(id, code, message)
    local response = {
        jsonrpc = "2.0",
        id      = id,
        error   = {
            code    = code,
            message = message,
        },
    }
    self:_send_message(response)
end

return JsonRpcClient

File: ruby/lib/copilot/json_rpc_client.rb
# frozen_string_literal: true

# Copyright (c) Microsoft Corporation. All rights reserved.

require "json"
require "securerandom"
require "thread"

module Copilot
  # JSON-RPC 2.0 error raised when the server returns an error response.
  class JsonRpcError < StandardError
    attr_reader :code, :data

    def initialize(code, message, data = nil)
      @code = code
      @data = data
      super("JSON-RPC Error #{code}: #{message}")
    end
  end

  # Minimal threaded JSON-RPC 2.0 client for stdio / TCP IO transport.
  #
  # Messages use Content-Length header framing (the LSP/JSON-RPC wire format):
  #   Content-Length: <byte-length>\r\n
  #   \r\n
  #   <JSON payload>
  #
  # The client runs a background reader thread that dispatches:
  # - Responses to pending requests (via per-request Queue)
  # - Notifications to the registered notification handler
  # - Incoming requests (server -> client) to registered request handlers
  class JsonRpcClient
    # @param input  [IO] readable IO (e.g. process stdout or TCP socket)
    # @param output [IO] writable IO (e.g. process stdin  or TCP socket)
    def initialize(input, output)
      @input  = input
      @output = output

      @pending_requests  = {}  # id => Queue
      @pending_lock      = Mutex.new

      @write_lock        = Mutex.new

      @notification_handler = nil  # Proc(method, params)
      @request_handlers     = {}   # method => Proc(params) -> Hash

      @running = false
      @reader_thread = nil
    end

    # Start the background reader thread.
    def start
      return if @running

      @running = true
      @reader_thread = Thread.new { read_loop }
      @reader_thread.abort_on_exception = false
    end

    # Stop the background reader thread.
    def stop
      @running = false
      @reader_thread&.join(2.0)

      # Unblock any pending requests so callers don't hang forever.
      @pending_lock.synchronize do
        @pending_requests.each_value do |queue|
          queue << { __error: JsonRpcError.new(-32000, "Client stopped") }
        end
        @pending_requests.clear
      end
    end

    # Send a JSON-RPC request and wait synchronously for the response.
    #
    # @param method  [String]       the RPC method name
    # @param params  [Hash, nil]    optional parameters
    # @param timeout [Numeric, nil] seconds to wait (default 30)
    # @return [Object] the +result+ from the JSON-RPC response
    # @raise [JsonRpcError] if the server returns an error
    # @raise [Timeout::Error] if the request times out
    def request(method, params = nil, timeout: 30, **kwargs)
      params = kwargs if params.nil? && !kwargs.empty?
      request_id = SecureRandom.uuid
      queue = Queue.new

      @pending_lock.synchronize do
        @pending_requests[request_id] = queue
      end

      send_message({
        jsonrpc: "2.0",
        id: request_id,
        method: method,
        params: params || {},
      })

      response = nil
      begin
        # Queue#pop with a timeout: use Timeout or poll.  We use a simple
        # polling approach to avoid Timeout's thread-safety issues.
        deadline = Time.now + (timeout || 30)
        loop do
          begin
            response = queue.pop(true) # non-blocking
            break
          rescue ThreadError
            # queue empty
            if Time.now > deadline
              raise Timeout::Error, "JSON-RPC request '#{method}' timed out after #{timeout}s"
            end
            sleep 0.01
          end
        end
      ensure
        @pending_lock.synchronize { @pending_requests.delete(request_id) }
      end

      if response.is_a?(Hash) && response.key?(:__error)
        raise response[:__error]
      end

      response
    end

    # Send a JSON-RPC notification (fire-and-forget, no response expected).
    #
    # @param method [String]    the RPC method name
    # @param params [Hash, nil] optional parameters
    def notify(method, params = nil, **kwargs)
      params = kwargs if params.nil? && !kwargs.empty?
      send_message({
        jsonrpc: "2.0",
        method: method,
        params: params || {},
      })
    end

    # Register a handler for incoming notifications from the server.
    # The handler receives +(method, params)+.
    #
    # @yield [method, params]
    def on_notification(&handler)
      @notification_handler = handler
    end

    # Register a handler for incoming requests from the server.
    # The handler receives +(params)+ and must return a Hash result.
    #
    # @param method [String] the RPC method to handle
    # @yield [params] the request parameters
    # @yieldreturn [Hash] the result to send back
    def on_request(method, &handler)
      if handler.nil?
        @request_handlers.delete(method)
      else
        @request_handlers[method] = handler
      end
    end

    private

    # Write a JSON-RPC message with Content-Length framing.
    def send_message(message)
      content = JSON.generate(message)
      content_bytes = content.encode("utf-8")
      header = "Content-Length: #{content_bytes.bytesize}\r\n\r\n"

      @write_lock.synchronize do
        @output.write(header)
        @output.write(content_bytes)
        @output.flush
      end
    end

    # Background loop: read messages and dispatch them.
    def read_loop
      while @running
        message = read_message
        break unless message

        handle_message(message)
      end
    rescue IOError, EOFError, Errno::EPIPE, Errno::ECONNRESET => _e
      # Stream closed; stop gracefully.
    rescue StandardError => e
      $stderr.puts("[CopilotSDK] JSON-RPC read loop error: #{e.class}: #{e.message}") if @running
    end

    # Read a single Content-Length framed JSON-RPC message (blocking).
    #
    # @return [Hash, nil] parsed message or nil on EOF
    def read_message
      # Read header line
      header_line = @input.gets
      return nil if header_line.nil?

      header = header_line.strip
      return nil unless header.start_with?("Content-Length:")

      content_length = header.split(":")[1].strip.to_i

      # Read the blank separator line
      @input.gets

      # Read exact content bytes, handling partial reads
      content_bytes = read_exact(content_length)
      return nil if content_bytes.nil?

      JSON.parse(content_bytes)
    end

    # Read exactly +num_bytes+ from the input, handling partial reads.
    #
    # @param num_bytes [Integer]
    # @return [String, nil]
    def read_exact(num_bytes)
      chunks = []
      remaining = num_bytes
      while remaining > 0
        chunk = @input.read(remaining)
        return nil if chunk.nil? || chunk.empty?

        chunks << chunk
        remaining -= chunk.bytesize
      end
      chunks.join
    end

    # Dispatch an incoming message to the appropriate handler.
    def handle_message(message)
      # Response to a pending request
      if message.key?("id") && (message.key?("result") || message.key?("error"))
        handle_response(message)
        return
      end

      # Notification from server (no id)
      if message.key?("method") && !message.key?("id")
        handle_notification(message)
        return
      end

      # Incoming request from server (has both method and id)
      if message.key?("method") && message.key?("id")
        handle_incoming_request(message)
        return
      end
    end

    # Resolve or reject a pending request.
    def handle_response(message)
      request_id = message["id"]
      queue = @pending_lock.synchronize { @pending_requests[request_id] }
      return unless queue

      if message.key?("error")
        err = message["error"]
        queue << {
          __error: JsonRpcError.new(
            err.fetch("code", -1),
            err.fetch("message", "Unknown error"),
            err["data"]
          )
        }
      elsif message.key?("result")
        queue << message["result"]
      else
        queue << { __error: JsonRpcError.new(-32600, "Invalid JSON-RPC response") }
      end
    end

    # Dispatch a notification to the registered handler.
    def handle_notification(message)
      return unless @notification_handler

      method = message["method"]
      params = message.fetch("params", {})
      @notification_handler.call(method, params)
    rescue StandardError => e
      $stderr.puts("[CopilotSDK] Notification handler error: #{e.class}: #{e.message}")
    end

    # Handle an incoming request from the server and send back a response.
    def handle_incoming_request(message)
      method = message["method"]
      request_id = message["id"]
      params = message.fetch("params", {})

      handler = @request_handlers[method]
      unless handler
        send_error_response(request_id, -32601, "Method not found: #{method}")
        return
      end

      # Execute handler (may be synchronous or may raise)
      begin
        result = handler.call(params)
        # sessionFs.* write-ops (writeFile/appendFile/mkdir/rm/rename) signal success
        # as JSON null (the CLI decodes the result as `SessionFsError | undefined`).
        # Coercing nil to {} would be decoded as a malformed SessionFsError
        # ("missing field `code`"), so preserve nil for those methods only.
        result = {} if result.nil? && !method.to_s.start_with?("sessionFs.")
        send_response(request_id, result)
      rescue JsonRpcError => e
        send_error_response(request_id, e.code, e.message, e.data)
      rescue StandardError => e
        send_error_response(request_id, -32603, e.message)
      end
    end

    # Send a successful response.
    def send_response(request_id, result)
      send_message({
        jsonrpc: "2.0",
        id: request_id,
        result: result,
      })
    end

    # Send an error response.
    def send_error_response(request_id, code, message, data = nil)
      error_obj = { code: code, message: message }
      error_obj[:data] = data if data
      send_message({
        jsonrpc: "2.0",
        id: request_id,
        error: error_obj,
      })
    end
  end
end
